README.md

```diff
@@ -16,85 +16,96 @@ Without needing to worry about instruction selection, register pressure, and so
-- To describe this pipeline, we need one variable `x`.
-- Both `func` objects have the same signature:
-	- Consume a buffer of `const int`, produce a buffer of `int`.
-	- The output dimension is indexed by `x`, and both operations require a the single point interval `[x, x]` of their inputs.
-	- `multiply_2` and `add_1` are functions implementing this operation.
+Without any further specifications this pipeline would use the following strategy:
 
-The possible implementations of this pipeline vary between two extremes:
-1. Allocating `intm` to have the same size as `out`, and executing all of `mul`, followed by all of `add`.
-2. Allocating `intm` to have a single element, and executing `mul` followed by `add` in a single loop over the output elements.
+- [1]  Materialize the whole logical `tmp` buffer of the same size as `out`, execute all of `mul`, followed by all of `add`.
 
-Of course, (2) would have extremely high overhead, and would not be a desireable strategy.
-If the buffers are large, (1) is inefficient due to poor memory locality.
-The ideal strategy is to split `out` into chunks, and compute the two operations at each chunk.
-This allows targeting a chunk size that fits in the cache, but amortizes the overhead of setting up the buffers and calling the functions implementing this operation.
-This can be implemented with the following schedule:
+This approach has multiple downsides: 1. large memory allocation for `tmp` 2.`tmp` memory is first fully written and then fully read, resulting
+in expensive memory traffic and inefficient use of CPU caches.
 
 ```c++
-const int chunk_size = 8;
-add.loops({x, chunk_size});
-mul.compute_at({&stencil, x});
+check(in)
+check((buffer_rank(in) == 1))
+check(out)
+check((buffer_rank(out) == 1))
+check((buffer_elem_size(in) == 4))
+check((buffer_elem_size(out) == 4))
+check(or_else((buffer_fold_factor(out, 0) == 9223372036854775807), (buffer_max(out, 0) < (buffer_fold_factor(out, 0) + buffer_min(out, 0)))))
+check((buffer_min(in, 0) <= buffer_min(out, 0)))
+check((buffer_max(out, 0) <= buffer_max(in, 0)))
+tmp = allocate(automatic, 4, {
+  {[buffer_min(out, 0), buffer_max(out, 0)], <>, <>}
+}) {
+ call(<mul>, {in}, {tmp})
+ call(<add>, {tmp}, {out})
+}
 ```
 
-In this case, the `mul.compute_at` specification is only for illustration purposes, it is equivalent to the default behavior, which is to compute funcs at the innermost location that does not imply redundant compute.
+The opposite extreme would be the following:
 
-This will result in a slinky program that looks like this:
+- [2] Allocate a buffer of size 1 for `tmp`. For each element compute `mul` followed by `add` and store the result to `out`.
+
+```c++
+...
+const int chunk_size = 1;
+add.loops({{x, chunk_size}});
+pipeline p = build_pipeline(ctx, {in}, {out});
+```
 
 ```c++
-intm = allocate(heap, 4, {
-  {[buffer_min(out, 0), buffer_max(out, 0)], 4, 8}
+tmp = allocate(automatic, 4, {
```

Comment:
We should update the other slinky programs in the README so they aren't using inconsistent syntax/keywords/etc.